# 01 - Data Understanding

## Objective

The objective of this notebook is to understand the structure, characteristics, and quality of the raw dataset before performing data preprocessing and feature engineering.

## Import libraries

In [6]:
import pandas as pd
import numpy as np

## Load dataset

In [7]:
df = pd.read_csv("../data/raw/student_dropout_dataset.csv")

In [8]:
df.head()

,Student_ID,Age,Gender,Family_Income,Internet_Access,Study_Hours_per_Day,Attendance_Rate,Assignment_Delay_Days,Travel_Time_Minutes,Part_Time_Job,Scholarship,Stress_Index,GPA,Semester_GPA,CGPA,Semester,Department,Parental_Education,Dropout
0,1,22.1,Male,25000.0,Yes,3.36,86.1,2,20.4,Yes,No,5.5,0.96,0.90,0.90,Year 1,Arts,High School,0
1,2,20.7,Male,25000.0,Yes,4.30,68.0,2,44.0,No,No,6.8,1.28,1.20,1.19,Year 3,Engineering,Bachelor,1
2,3,22.4,Male,40183.0,Yes,4.40,70.9,0,48.9,Yes,No,5.5,1.68,1.32,1.32,Year 1,Arts,Master,0
3,4,24.4,Male,NaN,Yes,NaN,82.2,2,38.6,No,No,NaN,1.78,1.77,1.77,Year 1,CS,High School,1
4,5,20.5,Female,25319.0,Yes,4.19,75.7,1,23.0,No,No,7.0,1.48,0.91,0.87,Year 4,Business,Bachelor,0


## Dataset dimensions

In [9]:
print("Number of records:", df.shape[0])
print("Number of features:", df.shape[1])

Number of records: 10000
Number of features: 19


## Understand columns

In [10]:
df.columns.tolist()

['Student_ID',
 'Age',
 'Gender',
 'Family_Income',
 'Internet_Access',
 'Study_Hours_per_Day',
 'Attendance_Rate',
 'Assignment_Delay_Days',
 'Travel_Time_Minutes',
 'Part_Time_Job',
 'Scholarship',
 'Stress_Index',
 'GPA',
 'Semester_GPA',
 'CGPA',
 'Semester',
 'Department',
 'Parental_Education',
 'Dropout']

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 19 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Student_ID             10000 non-null  int64  
 1   Age                    10000 non-null  float64
 2   Gender                 10000 non-null  object 
 3   Family_Income          9500 non-null   float64
 4   Internet_Access        10000 non-null  object 
 5   Study_Hours_per_Day    9500 non-null   float64
 6   Attendance_Rate        10000 non-null  float64
 7   Assignment_Delay_Days  10000 non-null  int64  
 8   Travel_Time_Minutes    10000 non-null  float64
 9   Part_Time_Job          10000 non-null  object 
 10  Scholarship            10000 non-null  object 
 11  Stress_Index           9500 non-null   float64
 12  GPA                    10000 non-null  float64
 13  Semester_GPA           10000 non-null  float64
 14  CGPA                   10000 non-null  float64
 15  Sem

## Target Variable

The target variable represents the outcome that the machine learning model will predict.

The target variable is `Dropout`.

The distribution of the target classes was examined to identify whether the dataset is balanced or imbalanced.

In [13]:
target_column = "Dropout"

df[target_column].value_counts()

Dropout
0    7646
1    2354
Name: count, dtype: int64

In [14]:
df[target_column].value_counts(normalize=True)

Dropout
0    0.7646
1    0.2354
Name: proportion, dtype: float64

## Missing Values

Missing values were examined for every feature.

Features with missing values will be handled during the preprocessing stage using an appropriate imputation strategy.

In [15]:
missing_values = df.isnull().sum()

missing_values[missing_values > 0]

Family_Income          500
Study_Hours_per_Day    500
Stress_Index           500
Parental_Education     511
dtype: int64

In [16]:
missing_percentage = (df.isnull().sum() / len(df)) * 100

missing_percentage[missing_percentage > 0]

Family_Income          5.00
Study_Hours_per_Day    5.00
Stress_Index           5.00
Parental_Education     5.11
dtype: float64

## Duplicate Records

In [17]:
print("Number of duplicate records:", df.duplicated().sum())

Number of duplicate records: 0


## Statistical Summary

The statistical summary was used to understand the distribution and range of numerical variables.

Some numerical features have considerably different scales, which may require feature scaling during preprocessing.

In [18]:
df.describe()

,Student_ID,Age,Family_Income,Study_Hours_per_Day,Attendance_Rate,Assignment_Delay_Days,Travel_Time_Minutes,Stress_Index,GPA,Semester_GPA,CGPA,Dropout
count,10000.00000,10000.00000,9500.000000,9500.000000,10000.00000,10000.000000,10000.00000,9500.000000,10000.000000,10000.000000,10000.000000,10000.00000
mean,5000.50000,21.02606,38377.247474,4.014592,81.73683,1.799700,30.17926,5.507147,2.308440,2.300057,2.298761,0.23540
std,2886.89568,2.13981,20496.232179,1.295450,8.22093,1.344307,11.91887,1.765951,1.061717,1.074407,1.072555,0.42427
min,1.00000,17.00000,25000.000000,0.500000,38.20000,0.000000,5.00000,1.000000,0.000000,0.000000,0.000000,0.00000
25%,2500.75000,19.50000,25000.000000,3.160000,76.40000,1.000000,21.90000,4.300000,1.550000,1.520000,1.520000,0.00000
50%,5000.50000,21.00000,29740.500000,4.000000,81.80000,2.000000,30.20000,5.500000,2.350000,2.350000,2.350000,0.00000
75%,7500.25000,22.50000,44520.000000,4.870000,87.30000,3.000000,38.40000,6.700000,3.120000,3.150000,3.150000,0.00000
max,10000.00000,29.60000,316601.000000,8.980000,100.00000,8.000000,74.90000,10.000000,4.000000,4.000000,4.000000,1.00000


In [19]:
df.describe(include="object")

,Gender,Internet_Access,Part_Time_Job,Scholarship,Semester,Department,Parental_Education
count,10000,10000,10000,10000,10000,10000,9489
unique,2,2,2,2,4,5,4
top,Female,Yes,No,No,Year 4,Science,Bachelor
freq,5011,8769,5996,6489,2536,2061,3949


## Categorical Features

In [20]:
categorical_columns = df.select_dtypes(include=["object"]).columns

categorical_columns

Index(['Gender', 'Internet_Access', 'Part_Time_Job', 'Scholarship', 'Semester',
       'Department', 'Parental_Education'],
      dtype='object')

In [21]:
for col in categorical_columns:
    print("\n", col)
    print(df[col].value_counts())


 Gender
Gender
Female    5011
Male      4989
Name: count, dtype: int64

 Internet_Access
Internet_Access
Yes    8769
No     1231
Name: count, dtype: int64

 Part_Time_Job
Part_Time_Job
No     5996
Yes    4004
Name: count, dtype: int64

 Scholarship
Scholarship
No     6489
Yes    3511
Name: count, dtype: int64

 Semester
Semester
Year 4    2536
Year 3    2521
Year 2    2488
Year 1    2455
Name: count, dtype: int64

 Department
Department
Science        2061
Arts           2026
Business       2002
CS             1974
Engineering    1937
Name: count, dtype: int64

 Parental_Education
Parental_Education
Bachelor       3949
High School    3072
Master         2003
PhD             465
Name: count, dtype: int64


## Numerical Features

In [22]:
numerical_columns = df.select_dtypes(include=np.number).columns

numerical_columns

Index(['Student_ID', 'Age', 'Family_Income', 'Study_Hours_per_Day',
       'Attendance_Rate', 'Assignment_Delay_Days', 'Travel_Time_Minutes',
       'Stress_Index', 'GPA', 'Semester_GPA', 'CGPA', 'Dropout'],
      dtype='object')

## Data Quality Issues

The following data quality issues were identified during the initial analysis:

* **Missing values:** `Family_Income`, `Study_Hours_per_Day`, `Stress_Index`, and `Parental_Education` contain approximately 5% missing values. These will be handled during preprocessing using appropriate imputation techniques.

* **Duplicate records:** No duplicate records were found.

* **Categorical values:** No inconsistent categorical values were identified.

* **Irrelevant features:** `Student_ID` is a unique identifier and has no predictive value, so it will be removed before model training.

* **Different feature scales:** Numerical features have different ranges. Feature scaling will therefore be considered during preprocessing for scale-sensitive models.

* **Potential outliers:** `Family_Income` shows a highly right-skewed distribution with potential high-value outliers. Other numerical features will also be examined for outliers during preprocessing.

* **Class imbalance:** The target variable `Dropout` is imbalanced, with 76.46% Not Dropout and 23.54% Dropout. Appropriate techniques such as class weighting or resampling will be considered during model development.
